src/inmanta/execute/dataflow/root_cause.py

```diff
@@ -72,31 +72,51 @@ def root_causes(self) -> Set[AttributeNode]:
             Rules 2 to 4 are implemented as propagation steps by
             _assignment_step, _child_attribute_step and _parent_instance_step respectively.
         """
-        causes: Set[AttributeNode] = set(())
-        for node in self.nodes:
-            others: FrozenSet[AttributeNode] = self.nodes.difference({node})
-            seen: Set[AssignableNode] = set(())
-            to_check: List[AssignableNode] = [node]
-
-            def process_step(step_result: FrozenSet[AssignableNode]) -> None:
-                new: Set[AssignableNode] = set(step_result).difference(seen)
-                to_check.extend(new)
-                seen.update(new)
-
-            is_root_cause: bool = True
-            while to_check:
-                n: AssignableNode = to_check.pop()
-                if n in others and n not in node.equivalence.nodes:
-                    is_root_cause = False
-                    break
-                if n.result_variable is not None and n.result_variable.hasValue:
-                    continue
-                process_step(self._assignment_step(n))
-                process_step(self._parent_instance_step(n))
-                process_step(self._child_attribute_step(n))
-            if is_root_cause:
-                causes.add(node)
-        return causes
+        roots: Set[AttributeNode] = set(())
+        seen: Set[AttributeNode] = set(())
+        to_check: List[AssignableNode] = self.nodes
+
+        def has_root(node: AssignableNode, cycledetect=set()) -> bool:
```

Comment:
```suggestion
        def has_root(node: AssignableNode, cycledetect: FrozenSet = frozenset()) -> bool:
```
Because it will always be the same object we should make sure it's immutable. This or default to None.